src/channel_manager.ts

```diff
@@ -0,0 +1,527 @@
+        notificationRemovedFromChannelHandler: this.notificationRemovedFromChannelHandler,
+      }) as [string, EventHandlerType<SCG>][],
+    );
+  }
+
+  public setChannels = (valueOrFactory: ChannelSetterParameterType<SCG>) => {
+    this.state.next((current) => {
+      const { channels: prevChannels } = current;
+      let newValue = valueOrFactory;
+      if (newValue && typeof newValue === 'function') {
+        newValue = newValue(prevChannels);
+      }
+
+      // If the references between the two values are the same, just return the
+      // current state; otherwise trigger a state change.
+      if (prevChannels === newValue) {
+        return current;
+      }
+      return { ...current, channels: newValue };
+    });
+  };
+
+  public setEventHandlerOverrides = (eventHandlerOverrides: ChannelManagerEventHandlerOverrides<SCG> = {}) => {
+    const truthyEventHandlerOverrides = Object.entries(eventHandlerOverrides).reduce<
+      Partial<ChannelManagerEventHandlerOverrides<SCG>>
+    >((acc, [key, value]) => {
+      if (value) {
+        acc[key as keyof ChannelManagerEventHandlerOverrides<SCG>] = value;
+      }
+      return acc;
+    }, {});
+    this.eventHandlerOverrides = new Map(
+      Object.entries(truthyEventHandlerOverrides) as [string, EventHandlerOverrideType<SCG>][],
+    );
+  };
+
+  public setOptions = (options: ChannelManagerOptions = {}) => {
+    this.options = { ...DEFAULT_CHANNEL_MANAGER_OPTIONS, ...options };
+  };
+
+  public queryChannels = async (
+    filters: ChannelFilters<SCG>,
+    sort: ChannelSort<SCG> = [],
+    options: ChannelOptions = {},
+    stateOptions: ChannelStateOptions = {},
+  ) => {
+    const { offset = 0, limit = 10 } = options;
+    const {
+      pagination: { isLoading },
+    } = this.state.getLatestValue();
+
+    if (isLoading) {
+      return;
+    }
+
+    try {
+      this.stateOptions = stateOptions;
+      this.state.next((currentState) => ({
+        ...currentState,
+        pagination: {
+          ...currentState.pagination,
+          isLoading: true,
+          isLoadingNext: false,
+          filters,
+          sort,
+          options,
+        },
+      }));
+
+      const channels = await this.client.queryChannels(filters, sort, options, stateOptions);
+      const newOffset = offset + (channels?.length ?? 0);
+      const newOptions = { ...options, offset: newOffset };
+      const { pagination } = this.state.getLatestValue();
+
+      this.state.partialNext({
+        channels,
+        pagination: {
+          ...pagination,
+          hasNext: (channels?.length ?? 0) >= limit,
+          isLoading: false,
+          options: newOptions,
+        },
+        ready: true,
+      });
+    } catch (error) {
+      this.client.logger('error', (error as Error).message);
+      this.state.next((currentState) => ({
+        ...currentState,
+        pagination: { ...currentState.pagination, isLoading: false },
+      }));
```

Comment:
This part should probably live in `finally` block.

Reply:
Not necessarily, due to 2 reasons: 

- It won't mean that `ready` will get set if the request failed
- `channels` will also not be set if the request failed

If we wanted to put it in a `finally` block, it would mean that we need to do 2 state updates whenever the query succeeds: 

- One to actually set the state (`ready`, `channels` etc)
- One to set `isLoading`

which I believe we want to avoid